File: src/imagebuilder/flyingtux/app/deploy.py
from ..services import process_os_service, get_os_service
from ..spec.deploy import DeploySpec, DeploySpec_representer
from ..naming import app_container_name
from metux.util.fs import mkdir
from os.path import isfile, dirname
from os import chmod
from toolbase import ToolBase
import yaml

class Deploy(ToolBase):

    def post_init(self):
        self.my_deploy_spec_file = self['TARGET::deploy-app-dir']+'/'+self['IMAGE::NAME']+'/info.yml'

    def run(self):
        self.create_deploy()
        self.create_script()

    def create_script(self):
        scriptdir = self['TARGET::workdir']+'/bin'
        name      = self['IMAGE::NAME']
        codebase  = self['TARGET::CODEBASE']
        scriptname = scriptdir+'/'+name
        mkdir(scriptdir)

        with open(scriptname, "w") as f:
            f.write('#!/usr/bin/python\n')
            f.write('import sys\n')
            f.write('sys.path.append("'+codebase+'")\n')
            f.write('from flyingtux.cmd import run_app\n')
            f.write('run_app()\n')
        chmod(scriptname, 0755)

    def create_deploy(self):
        image = self['IMAGE']
        self.info("deploying "+image['NAME'])
        mkdir(dirname(self.my_deploy_spec_file))

        # FIXME: autogenerate missing image on (re)deploy
        # FIXME: support transparent redeploy (eg. platform/arch changes)

        self.my_deploy_spec = DeploySpec({})
        d = self.my_deploy_spec
        if isfile(self.my_deploy_spec_file):
            self.info("already deployed: "+self.my_deploy_spec_file)
            self.my_deploy_spec.load_spec(self.my_deploy_spec_file)
        else:
            self.info("creating new deployment: "+self.my_deploy_spec_file)

        # write / update deployment descriptor
        d['image']         = image['NAME']
        d['name']          = app_container_name(appname=image['NAME'],
                                                version=image['version'])
        d['version']       = image['version']
        d['os-services']   = image['os-services']
        d['arch']          = self['ARCH']
        d['platform']      = self['PLATFORM::NAME']
        d['engine']        = self['TARGET::jail-engine']
        d['command']       = image['command']
        d['rootfs-image']  = self['ROOTFS-IMAGE']
        d['tmpdirs']       = (image.get_cf_list('rootfs::tmpdirs')
                            + image.get_cf_list('IMAGE::OSBASE::tmpdirs'))
        d['user']          = image['user']
        d['volumes']       = image['volumes']

        for sname,sspec in image['os-services'].iteritems():
            d['os-services::'+sname] = get_os_service(sname, sspec, self).get_conf()

        with open(self.my_deploy_spec_file, 'w') as outfile:
            yaml.dump(self.my_deploy_spec, outfile, default_flow_style=False, indent=4)

        return 0

File: internal/app/deploy.go
package app

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/metux/flyingtux/internal/fs"
	"github.com/metux/flyingtux/internal/services"
	"github.com/metux/flyingtux/internal/spec"
	"github.com/metux/flyingtux/internal/util"
)

// Deploy generates an app's deployment descriptor + launcher script (port of
// app.deploy.Deploy).
type Deploy struct {
	toolBase
	image *spec.SpecObj
}

// Run writes the deployment descriptor and the launcher script.
func (d *Deploy) Run() (int, error) {
	if err := d.createDeploy(); err != nil {
		return 0, err
	}
	if err := d.createScript(); err != nil {
		return 0, err
	}
	return 0, nil
}

func (d *Deploy) deployFile() string {
	return d.GetStr("TARGET::deploy-app-dir") + "/" + d.GetStr("IMAGE::NAME") + "/info.yml"
}

func (d *Deploy) createDeploy() error {
	image := d.image
	d.info("deploying " + image.GetStr("NAME"))
	deployFile := d.deployFile()
	if err := fs.Mkdir(filepath.Dir(deployFile)); err != nil {
		return err
	}

	dep := spec.NewRoot()
	if _, err := os.Stat(deployFile); err == nil {
		d.info("already deployed: " + deployFile)
		if existing, err := spec.Load(deployFile); err == nil {
			dep = existing
		}
	} else {
		d.info("creating new deployment: " + deployFile)
	}

	dep.SetStr("image", image.GetStr("NAME"))
	dep.SetStr("name", util.AppContainerName(image.GetStr("NAME"), image.GetStr("version")))
	dep.SetStr("version", image.GetStr("version"))
	dep.SetEntry("os-services", image.GetEntry("os-services"))
	dep.SetStr("arch", d.GetStr("ARCH"))
	dep.SetStr("platform", d.GetStr("PLATFORM::NAME"))
	dep.SetStr("engine", d.GetStr("TARGET::jail-engine"))
	dep.SetEntry("command", image.GetEntry("command"))
	dep.SetStr("rootfs-image", d.GetStr("ROOTFS-IMAGE"))

	// NOTE: the Python used 'IMAGE::OSBASE::tmpdirs' here, which resolves to
	// empty (image has no IMAGE key); replicated verbatim for parity.
	tmpdirs := append(image.GetStrList("rootfs::tmpdirs"),
		image.GetStrList("IMAGE::OSBASE::tmpdirs")...)
	for _, t := range tmpdirs {
		dep.AppendStr("tmpdirs", t)
	}

	dep.SetStr("user", image.GetStr("user"))
	dep.SetEntry("volumes", image.GetEntry("volumes"))

	for _, name := range image.SubKeys("os-services") {
		sspec := image.Sub(spec.Key("os-services::" + string(name)))
		svc, err := services.GetService(string(name), sspec, d)
		if err != nil {
			return err
		}
		conf, err := svc.GetConf()
		if err != nil {
			return err
		}
		dep.SetEntry(spec.Key("os-services::"+string(name)), conf.Entry)
	}

	return spec.Store(deployFile, dep, 0o644)
}

func (d *Deploy) createScript() error {
	scriptdir := d.GetStr("TARGET::workdir") + "/bin"
	name := d.GetStr("IMAGE::NAME")
	codebase := d.GetStr("TARGET::CODEBASE")
	scriptname := scriptdir + "/" + name
	if err := fs.Mkdir(scriptdir); err != nil {
		return err
	}
	script := fmt.Sprintf("#!/bin/sh\nexec %q run %q \"$@\"\n", codebase, name)
	if err := os.WriteFile(scriptname, []byte(script), 0o755); err != nil {
		return err
	}
	return os.Chmod(scriptname, 0o755)
}

// --- services.Runner (deploy only needs Get; the rest are unused in GetConf) ---

func (d *Deploy) Get(key string) string         { return d.GetStr(spec.Key(key)) }
func (d *Deploy) IPAddress() string             { return "" }
func (d *Deploy) StartWebProxy() (string, error) { return "", util.Error("web proxy not available at deploy time") }
